### DISCUSSION TITLE
Level 2 Service Education DRAFT

### DISCUSSION THREAD

**Discussion by zsamji** [bcgov/sso-keycloak#137] [Ideas]
To follow up on our[ Level 1 Service Education, ](https://github.com/bcgov/sso-keycloak/discussions/136) we want to equip you with the information needed to get a sample app up and running on your end

[Level 1 Link](https://github.com/bcgov/sso-keycloak/discussions/136)

**Tokens**
A Token is the strange lengthy string provided by the authentication server, which is actually used for authentication [Reference](https://medium.com/@danyalraja/oauth-in-plain-english-f1b0bc5275f0).  The frequently used ones are:

   - ID Token
      - ID tokens are issued by the authorization server and contain claims that carry information about the user. They can be sent alongside or instead of an access token. Information in ID Tokens allows the client to verify that a user is who they claim to be. [Reference](https://learn.microsoft.com/en-us/azure/active-directory/develop/id-tokens)
   - Access Token
      -  An access token is a tiny piece of code that contains a large amount of data. Information about the user, permissions, groups, and timeframes is embedded within one token that passes from a server to a user's device. Plenty of websites use access tokens. [Reference](https://www.okta.com/identity-101/access-token/)

   - [What's the difference ID Token and Access Token](https://auth0.com/blog/id-token-access-token-what-is-the-difference/)

   - Refresh Token
      - A refresh token is a special key that enables a client for an API or service to retrieve new access tokens without requiring the user to perform a complete login. In other words, an application can exchange a valid refresh token for a new access token. [Reference](https://passage.id/post/how-refresh-tokens-work-a-complete-guide-for-beginners)
      - 
[Refresh Tokens: What and when](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/)

More on tokens
- [What are JSON Web Tokens (JWT), JSON Web Key (JWK), etc](https://codecurated.com/blog/introduction-to-jwt-jws-jwe-jwa-jwk/) 

   - 
**Token Validation - why you care and where you do this**
- Where you can validate tokens (front end or back end) 
   - [JavaScript Object Notation - JSON Web tokens](https://auth0.com/learn/json-web-tokens/)
   - [What to check when validating ID token](https://developer.okta.com/docs/guides/validate-id-tokens/main/#what-to-check-when-validating-an-id-token)
   - [Validating access tokens](https://www.keycloak.org/docs/latest/securing_apps/#validating-access-tokens)

**What do tokens give you?**
 - What the tokens give you ie attributes or mappers (specific identity information)
   - [Visual Representation](https://github.com/bcgov/sso-keycloak/wiki/Identity-Provider-Attribute-Mapping)
   - [Another representation of IDP Attributes & Mappers](https://github.com/bcgov/sso-keycloak/blob/dev/docs/migration-guide.md)
   - [What do I need, Public or Confidential Client Type?]( https://github.com/bcgov/sso-keycloak/wiki/Using-Your-SSO-Client#confidential-vs-public-client)
- Use our [Common Hosted Single Sign On  (CSS) App to make an integration request: ](https://bcgov.github.io/sso-requests)

**Examples you can use**
    - [Example Repo](https://github.com/bcgov/keycloak-example-apps/tree/dev/examples)
    
Stay tuned as we create a new discussion thread on Level 3 that dives deeper into other things....

**Comment by kurokn** [comment]
Great links in defining the different types of tokens, not sure if reader needs to understand all types, but it's good to have them. Token samples are very useful and in a fair level of difficulty in understanding them. Personally feel the token validation part is the hardest one on this level.

**Reply by zsamji** [reply]
i've just added some terms on tokens - take a peak; i still need to add a definition of payload and and token validation
